src/compute-client/src/controller/instance.rs

```diff
@@ -1800,40 +1739,29 @@ where
             // Apply read capability updates and learn about resulting changes to the read
             // frontier.
             let changes = collection.read_capabilities.update_iter(update.drain());
-            update.extend(changes);
-
-            if update.is_empty() {
+            if changes.count() == 0 {
                 continue; // read frontier did not change
             }
 
-            compute_downgraded.insert(id);
+            let new_since = collection.read_frontier().to_owned();
 
-            // Propagate read frontier updates to dependencies.
-            for dep_id in &collection.compute_dependencies {
-                updates
-                    .entry(*dep_id)
-                    .or_insert_with(ChangeBatch::new)
-                    .extend(update.iter().cloned());
+            // Propagate read frontier update to dependencies.
+            for read_hold in collection.compute_dependencies.values_mut() {
+                read_hold
+                    .try_downgrade(new_since.clone())
+                    .expect("frontiers don't regress");
             }
-            for dep_id in &collection.storage_dependencies {
-                storage_updates
-                    .entry(*dep_id)
-                    .or_insert_with(ChangeBatch::new)
-                    .extend(update.iter().cloned());
+            for read_hold in collection.storage_dependencies.values_mut() {
+                read_hold
+                    .try_downgrade(new_since.clone())
+                    .expect("frontiers don't regress");
             }
```

Comment:
Note that these downgrades only get picked up the next time `apply_read_hold_changes` is invoked, so dependency downgrades now propagate slower than before. A later commit fixes that by inlining `update_read_capabilities` into `apply_read_hold_changes`.